File: src/main/java/hiiragi283/material/api/capability/fluid/HiiragiFluidTankWrapper.java
package hiiragi283.material.api.capability.fluid;

import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.nbt.NBTTagList;
import net.minecraftforge.common.util.Constants;
import net.minecraftforge.common.util.INBTSerializable;
import net.minecraftforge.fluids.FluidStack;
import net.minecraftforge.fluids.capability.IFluidHandler;
import net.minecraftforge.fluids.capability.IFluidTankProperties;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.annotation.ParametersAreNonnullByDefault;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;

@ParametersAreNonnullByDefault
public class HiiragiFluidTankWrapper implements IFluidHandler, INBTSerializable<NBTTagCompound> {

    @NotNull
    private List<HiiragiFluidTank> tanks;

    public void setSize(int size) {
        tanks = new ArrayList<>(size);
    }

    public HiiragiFluidTankWrapper(Collection<HiiragiFluidTank> tanks) {
        this.tanks = new ArrayList<>(tanks);
    }

    public HiiragiFluidTankWrapper(HiiragiFluidTank... tanks) {
        this.tanks = Arrays.asList(tanks);
    }

    //    FluidHandler    //

    @Override
    public IFluidTankProperties[] getTankProperties() {
        return tanks.stream().map(tank -> tank.getTankProperties()[0]).toArray(IFluidTankProperties[]::new);
    }

    @Override
    public int fill(FluidStack resource, boolean doFill) {
        int result = 0;
        for (HiiragiFluidTank tank : tanks) {
            if (tank.getIOType().canInsert) {
                int resultTest = tank.fill(resource, false);
                if (resultTest > 0) {
                    result = tank.fill(resource, doFill);
                    break;
                }
            }
        }
        return result;
    }

    @Nullable
    @Override
    public FluidStack drain(FluidStack resource, boolean doDrain) {
        FluidStack result = null;
        for (HiiragiFluidTank tank : tanks) {
            if (tank.getIOType().canExtract) {
                FluidStack resultTest = tank.drain(resource, false);
                if (resultTest != null) {
                    result = tank.drain(resource, doDrain);
                    break;
                }
            }
        }
        return result;
    }

    @Nullable
    @Override
    public FluidStack drain(int maxDrain, boolean doDrain) {
        FluidStack result = null;
        for (HiiragiFluidTank tank : tanks) {
            if (tank.getIOType().canExtract) {
                FluidStack resultTest = tank.drain(maxDrain, false);
                if (resultTest != null) {
                    result = tank.drain(maxDrain, doDrain);
                    break;
                }
            }
        }
        return result;
    }

    //    INBTSerializable<NBTTagCompound>    //

    @Override
    public NBTTagCompound serializeNBT() {
        var tagList = new NBTTagList();
        for (HiiragiFluidTank tank : tanks) {
            tagList.appendTag(tank.serializeNBT());
        }
        var tag = new NBTTagCompound();
        tag.setTag("Fluids", tagList);
        return tag;
    }

    @Override
    public void deserializeNBT(NBTTagCompound nbt) {
        NBTTagList tagList = nbt.getTagList("Fluids", Constants.NBT.TAG_COMPOUND);
        setSize(tagList.tagCount());
        for (int i = 0; i < tagList.tagCount(); i++) {
            NBTTagCompound tag = tagList.getCompoundTagAt(i);
            tanks.get(i).deserializeNBT(tag);
        }
    }

}

File: src/main/kotlin/hiiragi283/material/api/capability/fluid/HiiragiFluidTankWrapper.kt
package hiiragi283.material.api.capability.fluid

import net.minecraft.nbt.NBTTagCompound
import net.minecraft.nbt.NBTTagList
import net.minecraftforge.common.util.Constants
import net.minecraftforge.common.util.INBTSerializable
import net.minecraftforge.fluids.FluidStack
import net.minecraftforge.fluids.capability.IFluidHandler
import net.minecraftforge.fluids.capability.IFluidTankProperties

class HiiragiFluidTankWrapper(private val tanks: MutableList<HiiragiFluidTank>) : IFluidHandler,
    INBTSerializable<NBTTagCompound> {

    constructor(vararg tanks: HiiragiFluidTank) : this(tanks.toMutableList())

    //    FluidHandler    //

    override fun getTankProperties(): Array<IFluidTankProperties> {
        return tanks.flatMap { tank: HiiragiFluidTank -> tank.tankProperties.toList() }.toTypedArray()
    }

    override fun fill(resource: FluidStack, doFill: Boolean): Int {
        var result = 0
        for (tank: HiiragiFluidTank in tanks) {
            if (tank.ioType.canInsert) {
                val resultTest: Int = tank.fill(resource, false)
                if (resultTest > 0) {
                    result = tank.fill(resource, doFill)
                    break
                }
            }
        }
        return result
    }

    override fun drain(resource: FluidStack, doDrain: Boolean): FluidStack? = drain(resource.amount, doDrain)

    override fun drain(maxDrain: Int, doDrain: Boolean): FluidStack? {
        var result: FluidStack? = null
        for (tank: HiiragiFluidTank in tanks) {
            if (tank.ioType.canExtract) {
                val resultTest: FluidStack? = tank.drain(maxDrain, false)
                if (resultTest != null) {
                    result = tank.drain(maxDrain, doDrain)
                    break
                }
            }
        }
        return result
    }

    //    INBTSerializable<NBTTagCompound>    //

    override fun serializeNBT(): NBTTagCompound {
        val tagList = NBTTagList()
        tanks.forEach { tank: HiiragiFluidTank -> tagList.appendTag(tank.serializeNBT()) }
        return NBTTagCompound().also { tag: NBTTagCompound -> tag.setTag("Fluids", tagList) }
    }

    override fun deserializeNBT(nbt: NBTTagCompound) {
        val tagList: NBTTagList = nbt.getTagList("Fluids", Constants.NBT.TAG_COMPOUND)
        for (i in 0 until tagList.tagCount()) {
            tanks[i] = HiiragiFluidTank(0)
            tanks[i].deserializeNBT(tagList.getCompoundTagAt(i))
        }
    }

}